drivers/ethernet/eth_stm32_hal.c

```diff
@@ -221,17 +222,37 @@ static int eth_tx(struct device *dev, struct net_pkt *pkt)
 	tx_config.Length = total_len;
 	tx_config.TxBuffer = tx_buffer_def;
 
-	hal_ret = HAL_ETH_Transmit(heth, &tx_config, ETH_DMA_TRANSMIT_TIMEOUT);
+	/* tx_buffer is allocated on function stack, we need to wait for the transfer to complete */
+	/* So it is not freed before the interrupt happens */
+	hal_ret = HAL_ETH_Transmit_IT(heth, &tx_config);
+
+	if (hal_ret != HAL_OK) {
```

Comment:
Putting it at line 253 ( after `#endif /* CONFIG_SOC_SERIES_STM32H7X */`) would be better. You can use then this error check for both H7 and non H7 versions of the driver. It is also better I think to put this error after next `k_sem_take` since taking semaphore is more important than logging error here!

Reply:
If for some reason the start of TX transmission fails, then we mustn't block on the ```k_sem_take``` call since the semaphore will never be given due to the absence of TX complete interrupt. This would lock the function forever and you would never exit from it. Therefore I dont agree with this proposition 🙃

Reply:
Ok. You're right. Please mark this conversation as resolved then.